File: CS/GoogleCalendarExample/Form1.cs
using DevExpress.XtraScheduler;
using Google.Apis.Auth.OAuth2;
using Google.Apis.Calendar.v3;
using Google.Apis.Calendar.v3.Data;
using Google.Apis.Services;
using Google.Apis.Util.Store;
using System;
using System.Collections.Generic;
using System.IO;
using System.Threading;
using System.Windows.Forms;

namespace GoogleCalendarExample {
    public partial class Form1 : Form {
        const string DefaultCaption = "Google Calendar Importer";
        static string[] Scopes = { CalendarService.Scope.Calendar };
        static string ApplicationName = "GoogleCalendarExample";
        bool isConnected = false;
        public Form1() {
            InitializeComponent();
            this.schedulerControl1.Storage.Appointments.CustomFieldMappings.Add(new AppointmentCustomFieldMapping("eventId", "eventId"));
            GoogleCalendarExample.Log.Register(Log);
            UpdateFormState();

            this.schedulerControl1.Storage.AppointmentsDeleted += OnAppointmentsChanged;
            this.schedulerControl1.Storage.AppointmentsChanged += OnAppointmentsChanged;
            this.schedulerControl1.Storage.AppointmentsInserted += OnAppointmentsChanged;
        }
               

        CalendarService CalendarService { get; set; }
        bool LockStorageEvents { get; set; }

        void OnAppointmentsChanged(object sender, PersistentObjectsEventArgs e) {
            if (LockStorageEvents || !isConnected)
                return;
            CalendarListEntry calendarEntry = this.cbCalendars.SelectedValue as CalendarListEntry;
            CalendarExporter exporter = new CalendarExporter(CalendarService, calendarEntry);
            exporter.Export(e.Objects as IList<Appointment>);
        }

        void OnBtnConnectClick(object sender, EventArgs e) {
            UserCredential credential;
            using (var stream =
                new FileStream("secret\\client_secret.json", FileMode.Open, FileAccess.Read)) {
                string credPath = Environment.CurrentDirectory;
                credPath = Path.Combine(credPath, ".credentials");

                credential = GoogleWebAuthorizationBroker.AuthorizeAsync(
                    GoogleClientSecrets.Load(stream).Secrets,
                    Scopes,
                    "user",
                    CancellationToken.None,
                    new FileDataStore(credPath, true)).Result;
                Log("Credential file saved to: " + credPath);
            }
            isConnected = true;
            // Create Google Calendar API service.
            CalendarService = new CalendarService(new BaseClientService.Initializer() {
                HttpClientInitializer = credential,
                ApplicationName = ApplicationName,
            });
            var calendarListRequst = CalendarService.CalendarList.List();
            CalendarList calendarList = calendarListRequst.Execute();
            foreach (var item in calendarList.Items) {
                Log(item.Summary);
            }
            cbCalendars.DisplayMember = "Summary";
            cbCalendars.DataSource = calendarList.Items;

            cbCalendars.SelectedValueChanged += OnCbCalendarsSelectedValueChanged;
            UpdateFromGoogleCalendar();
        }
        void OnCbCalendarsSelectedValueChanged(object sender, EventArgs e) {
            UpdateFromGoogleCalendar();
        }

        void UpdateFromGoogleCalendar() {
            LockStorageEvents = true;

            CalendarListEntry calendarEntry = this.cbCalendars.SelectedValue as CalendarListEntry;
            Calendar calendar = CalendarService.Calendars.Get(calendarEntry.Id).Execute();
            EventsResource.ListRequest listRequest = CalendarService.Events.List(calendarEntry.Id);
            listRequest.MaxResults = 10000;
            Events events = listRequest.Execute();
            Log("Loaded {0} events", events.Items.Count);
            this.schedulerStorage1.Appointments.Items.Clear();
            this.schedulerStorage1.BeginUpdate();
            try {
                CalendarImporter importer = new CalendarImporter(this.schedulerStorage1);
                importer.Import(events.Items);
            } finally {
                this.schedulerStorage1.EndUpdate();
            }
            SetStatus(String.Format("Loaded {0} events", events.Items.Count));
            UpdateFormState();

            LockStorageEvents = false;
        }

        void UpdateFormState() {
            if (CalendarService == null) {
                this.cbCalendars.Enabled = false;
                this.btnConnect.Enabled = true;
            } else {
                this.cbCalendars.Enabled = true;
                this.btnConnect.Enabled = false;
            }
        }
                
        void OnBtnRefrehsClick(object sender, EventArgs e) {
            UpdateFromGoogleCalendar();
        }

        #region Logging
        void Log(string message) {
            this.tbLog.AppendText(message + "\r\n");
        }
        void Log(string format, params object[] args) {
            Log(string.Format(format, args));
        }
        void SetStatus(string message) {
            this.tsStatus.Text = message;
        }
        #endregion

    }
}

File: VB/GoogleCalendarExample/Form1.vb
Imports Microsoft.VisualBasic
Imports DevExpress.XtraScheduler
Imports Google.Apis.Auth.OAuth2
Imports Google.Apis.Calendar.v3
Imports Google.Apis.Calendar.v3.Data
Imports Google.Apis.Services
Imports Google.Apis.Util.Store
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Threading
Imports System.Windows.Forms

Namespace GoogleCalendarExample
	Partial Public Class Form1
		Inherits Form
		Private Const DefaultCaption As String = "Google Calendar Importer"
		Private Shared Scopes() As String = { CalendarService.Scope.Calendar }
		Private Shared ApplicationName As String = "GoogleCalendarExample"
		Private isConnected As Boolean = False
		Public Sub New()
			InitializeComponent()
			Me.schedulerControl1.Storage.Appointments.CustomFieldMappings.Add(New AppointmentCustomFieldMapping("eventId", "eventId"))
			GoogleCalendarExample.Log.Register(AddressOf Log)
			UpdateFormState()

			AddHandler Me.schedulerControl1.Storage.AppointmentsDeleted, AddressOf OnAppointmentsChanged
			AddHandler Me.schedulerControl1.Storage.AppointmentsChanged, AddressOf OnAppointmentsChanged
			AddHandler Me.schedulerControl1.Storage.AppointmentsInserted, AddressOf OnAppointmentsChanged
		End Sub


		Private privateCalendarService As CalendarService
		Private Property CalendarService() As CalendarService
			Get
				Return privateCalendarService
			End Get
			Set(ByVal value As CalendarService)
				privateCalendarService = value
			End Set
		End Property
		Private privateLockStorageEvents As Boolean
		Private Property LockStorageEvents() As Boolean
			Get
				Return privateLockStorageEvents
			End Get
			Set(ByVal value As Boolean)
				privateLockStorageEvents = value
			End Set
		End Property

		Private Sub OnAppointmentsChanged(ByVal sender As Object, ByVal e As PersistentObjectsEventArgs)
			If LockStorageEvents OrElse (Not isConnected) Then
				Return
			End If
			Dim calendarEntry As CalendarListEntry = TryCast(Me.cbCalendars.SelectedValue, CalendarListEntry)
			Dim exporter As New CalendarExporter(CalendarService, calendarEntry)
			exporter.Export(TryCast(e.Objects, IList(Of Appointment)))
		End Sub

		Private Sub OnBtnConnectClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnConnect.Click
			Dim credential As UserCredential
			Using stream = New FileStream("secret\client_secret.json", FileMode.Open, FileAccess.Read)
				Dim credPath As String = Environment.CurrentDirectory
				credPath = Path.Combine(credPath, ".credentials")

				credential = GoogleWebAuthorizationBroker.AuthorizeAsync(GoogleClientSecrets.Load(stream).Secrets, Scopes, "user", CancellationToken.None, New FileDataStore(credPath, True)).Result
				Log("Credential file saved to: " & credPath)
			End Using
			isConnected = True
			' Create Google Calendar API service.
			CalendarService = New CalendarService(New BaseClientService.Initializer() With {.HttpClientInitializer = credential, .ApplicationName = ApplicationName})
			Dim calendarListRequst = CalendarService.CalendarList.List()
			Dim calendarList As CalendarList = calendarListRequst.Execute()
			For Each item In calendarList.Items
				Log(item.Summary)
			Next item
			cbCalendars.DisplayMember = "Summary"
			cbCalendars.DataSource = calendarList.Items

			AddHandler cbCalendars.SelectedValueChanged, AddressOf OnCbCalendarsSelectedValueChanged
			UpdateFromGoogleCalendar()
		End Sub
		Private Sub OnCbCalendarsSelectedValueChanged(ByVal sender As Object, ByVal e As EventArgs)
			UpdateFromGoogleCalendar()
		End Sub

		Private Sub UpdateFromGoogleCalendar()
			LockStorageEvents = True

			Dim calendarEntry As CalendarListEntry = TryCast(Me.cbCalendars.SelectedValue, CalendarListEntry)
			Dim calendar As Calendar = CalendarService.Calendars.Get(calendarEntry.Id).Execute()
			Dim listRequest As EventsResource.ListRequest = CalendarService.Events.List(calendarEntry.Id)
			listRequest.MaxResults = 10000
			Dim events As Events = listRequest.Execute()
			Log("Loaded {0} events", events.Items.Count)
			Me.schedulerStorage1.Appointments.Items.Clear()
			Me.schedulerStorage1.BeginUpdate()
			Try
				Dim importer As New CalendarImporter(Me.schedulerStorage1)
				importer.Import(events.Items)
			Finally
				Me.schedulerStorage1.EndUpdate()
			End Try
			SetStatus(String.Format("Loaded {0} events", events.Items.Count))
			UpdateFormState()

			LockStorageEvents = False
		End Sub

		Private Sub UpdateFormState()
			If CalendarService Is Nothing Then
				Me.cbCalendars.Enabled = False
				Me.btnConnect.Enabled = True
			Else
				Me.cbCalendars.Enabled = True
				Me.btnConnect.Enabled = False
			End If
		End Sub

		Private Sub OnBtnRefrehsClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnRefresh.Click
			UpdateFromGoogleCalendar()
		End Sub

		#Region "Logging"
		Private Sub Log(ByVal message As String)
			Me.tbLog.AppendText(message & Constants.vbCrLf)
		End Sub
		Private Sub Log(ByVal format As String, ParamArray ByVal args() As Object)
			Log(String.Format(format, args))
		End Sub
		Private Sub SetStatus(ByVal message As String)
			Me.tsStatus.Text = message
		End Sub
		#End Region

	End Class
End Namespace
